File: src/java/engine/board/Piece.java
package java.engine.board;

/**
 * Created by melvic on 6/27/18.
 */
public interface Piece {
    public enum Type {
        PAWN, KNIGHT, BISHOP, ROOK, QUEEN, KING;
    }
    public enum Side {
        WHITE, BLACK;

        public Side opposite() {
            return values()[ordinal() ^ 1];
        }
    }

    public Type getType();
    public Side getSide();

    class Builder {
        private Side side;

        Builder(Side side) {
            this.side = side;
        }

        public Piece of(Type type) {
            return new Piece() {
                @Override
                public Type getType() {
                    return type;
                }

                @Override
                public Side getSide() {
                    return side;
                }

                @Override
                public String toString() {
                    return getSide() + " " + getType();
                }

                @Override
                public boolean equals(Object obj) {
                    if (!(obj instanceof Piece)) return false;
                    Piece that = (Piece) obj;
                    return this.getType() == that.getType()
                            && this.getSide() == that.getSide();
                }
            };
        }
    }

    public static Builder build(Side side) {
        return new Builder(side);
    }
    public static Builder white = new Builder(Side.WHITE);
    public static Builder black = new Builder(Side.BLACK);
}

File: src/scala/engine/board/Piece.scala
package scala.engine.board

/**
  * Created by melvic on 8/6/18.
  */
sealed trait PieceType
case object Pawn extends PieceType
case object Knight extends PieceType
case object Bishop extends PieceType
case object Rook extends PieceType
case object Queen extends PieceType
case object King extends PieceType

sealed trait Side {
  def opposite: Side = this match {
    case White => Black
    case Black => White
  }
}

case object White extends Side
case object Black extends Side

case class Piece(pieceType: PieceType, side: Side)

object Piece {
  def whiteOf(pieceType: PieceType) = Piece(pieceType, White)
  def blackOf(pieceType: PieceType) = Piece(pieceType, Black)
}
